### DISCUSSION TITLE
Extend captured variables from patterns to assign to existing variables

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#7001] [Language Ideas]
Currently you cannot directly assign a captured variable within a pattern to an existing local variable. Instead of conditionally assigning directly to the existing variable, you are required to use the conditionally assigned variable within an if statement, and manually assign it to the variable of your choice. This means that you add one extra variable, and introduce a new if statement in your code for something intuitive, such as exporting a result from a pattern match.

For example, I have to write this
```csharp
public static bool TryGetSingleSubdirectory(this DirectoryInfo directoryInfo, out DirectoryInfo? singleDirectory)
{
    singleDirectory = null;
    var infos = directoryInfo.GetFileSystemInfos();
    if (infos is [DirectoryInfo single])
    {
        singleDirectory = single;
    }
    return singleDirectory is not null;
}
```

To my mind, this is too much for something as simple as getting the result of the list pattern directly into the out parameter. Skipping the unavailability of the default out parameter value (see also #1033), I would imagine being able to write the above as:

```csharp
public static bool TryGetSingleSubdirectory(this DirectoryInfo directoryInfo, out DirectoryInfo? singleDirectory)
{
    singleDirectory = null;
    return directoryInfo.GetFileSystemInfos() is [out singleDirectory];
}
```

This one conditionally assigns the `singleDirectory` variable to the value captured by the expression, upcast into `DirectoryInfo` via `as`.
The simplified code should be equivalent to the example shown above.

The `out` keyword was chosen to include so as to avoid the usage of a conditional keyword that could be shadowed by a type identifier, disabling the feature. Also, passing a simple identifier would be parsed as a type, meaning there should be a way to indicate that a variable is being targeted.

Likewise, this proposal would extend to other type patterns, like for instance:
```csharp
public static bool IsDirectory(this FileSystemInfo? info, out DirectoryInfo? directory)
{
    DirectoryInfo? directory = null;
    return directory is out directory;
}
```

**Comment by CyrusNajmabadi** [comment]
I really want a way for a pattern to directly assign into an out variable. I'll champion this.

**Reply by En3Tho** [reply]
Do you want to scope this to out variable assignment only or introduce a general existing variable (re)assignment mechanism?

In different patterns you can't have variables with same names although scopes of those patterns of not intervene.

General mechanism can help with this (I personally would like to have compiler allow same name-different scope scenarios).

**Reply by 333fred** [reply]
@CyrusNajmabadi note that this is highly related to https://github.com/dotnet/csharplang/issues/4018, and LDM has extensively discussed this type of syntax before, with general negative sentiment.

**Reply by CyrusNajmabadi** [reply]
Understood.  but i would like *some* solution.  There seems to be a general pain point merging out-parameters and pattern matching that seems unnecessarily unpleasant.  :)

**Reply by 333fred** [reply]
That's fine, I was just reminding you of the previous conversations in case you'd forgotten.
